pkg/cloud/services/compute/instance.go

```diff
@@ -407,68 +386,129 @@ func (s *Service) getVolumeByName(name string) (*volumes.Volume, error) {
 	return &volumeList[0], nil
 }
 
-func (s *Service) getOrCreateRootVolume(eventObject runtime.Object, instanceSpec *InstanceSpec, imageID string) (*volumes.Volume, error) {
-	rootVolume := instanceSpec.RootVolume
-	if !hasRootVolume(rootVolume) {
-		return nil, nil
+func (s *Service) getOrCreateVolume(eventObject runtime.Object, opts volumes.CreateOpts) (*volumes.Volume, error) {
+	existingVolume, err := s.getVolumeByName(opts.Name)
+	if err != nil {
+		return nil, err
 	}
+	if existingVolume != nil {
+		if existingVolume.Size != opts.Size {
```

Comment:
Don't we need to also check the availability zone and the type to ensure the existing volume conforms to the spec?
Alternatively, it might be simpler to error out when there is an existing volume with the name we want to use.

Reply:
What if we do this:
* When we create a volume, we put the Kubernetes UUIDs of the OpenStackCluster and the OpenStackMachine in the volume metadata
  * openstack.infrastructure.cluster.x-k8s.io/cluster-uuid
  * openstack.infrastructure.cluster.x-k8s.io/machine-uuid
* When we search for a volume by name and find one, we then check if the UUIDs in the metadata match the cluster/machine being reconciled
* If they do, we own the volume and that is the volume we return
* If they don't, we don't own the volume and we error out

If that's a pattern people agree on, I'll be happy to do this via a separated PR if possible. For the sake of not adding too many design changes into one PR.

Reply:
@mdbooth and I also discussed doing this for Neutron resources using tags

Reply:
That looks reasonable. I would just change the logic to search for volume by name and by metadata in a single query. If it returns a result, that's the one we own and we can use it, otherwise, we know we need to create a new one.

Reply:
Not sure the single query is possible unfortunately, although I might have missed something. I would be surprised if the metadata items were indexed for searching.

Reply:
I think it's possible:

https://github.com/gophercloud/gophercloud/blob/1ef3b0843934066d5c65163e64683657608e6975/openstack/blockstorage/v3/volumes/requests.go#L125-L126

Reply:
We came to the agreement that we (I) will take care of it in a separated PR.

Reply:
Like I said on another comment, I'm working on it and this will be done before the next release is cut.